pkg/hubble/relay/pool/client_test.go

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package pool
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"math/big"
+	"net"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/health"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+
+	"github.com/cilium/cilium/pkg/crypto/certloader"
+	hubbleopts "github.com/cilium/cilium/pkg/hubble/server/serveroption"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+func TestGRPCClientConnBuilder_CertificateChange(t *testing.T) {
+	cert, ca := newTestCAandCert(t)
+
+	fTLSb := &fakeTLSConfigBuilder{
+		cert: &cert,
+		ca:   ca,
+	}
+	cb := GRPCClientConnBuilder{
+		DialTimeout: 5 * time.Second,
+		Options: []grpc.DialOption{
+			grpc.WithBlock(),
+			grpc.FailOnNonTempDialError(true),
+			grpc.WithReturnConnectionError(),
+		},
+		TLSConfig: fTLSb,
+	}
+
+	list, err := net.Listen("tcp", "localhost:0")
+	require.NoError(t, err)
+	addr := list.Addr().String()
+
+	s := newTestServer(cert, ca)
+	go s.Serve(list)
+
+	clientConn, err := cb.ClientConn(list.Addr().String(), "foo.test.cilium.io")
+	require.NoError(t, err)
+	hc := healthpb.NewHealthClient(clientConn)
+	_, err = hc.Check(context.TODO(), nil)
+	require.NoError(t, err)
+
+	cert, ca = newTestCAandCert(t)
+
+	s.Stop()
+	// Start server with new cert on same port
+	list, err = net.Listen("tcp", addr)
```

Comment:
This is technically a race condition, but I don't know how to fix it. I'm open to any idea on how to fix this test.
The issue is, that after closing and starting to listen again, someone else could start listening on the same port, which would cause the test to fail.

I tried to:

* Bring gRPC not to close the listener when stopping the server, but that doesn't seem to be possible
* Wrap the listener and do a noop on closing, but that turns into concurrency hell because we need to stop the `Accept` on `Close`. But there is now way to cancel the `Accept` so we need to handle this in a goroutine and coordinate between the accepting goroutine and the different `Accept` and `Close` calls and this is where I gave up.

We might be able to do something with `SO_REUSEPORT`, but that also seems overly complicated.

Reply:
Can you just let it assign a random port and connect to it? That's how the httptest server works IIRC.

Reply:
That's what we do for the first listen, but we need to use the same port again, when restarting the server with new certs.

So for the test the server starts listening on port A (which is a random free port). Client then connects to port A. We stop the server and start listening again on port A with a new certificate. The client then needs to be able to automatically update certificates and reconnect, without having to re-dial the gRPC ClientConn.

Reply:
We do need to stop the server. If we just reload the certificate it will keep old connections open for quite a while, so we need to stop the server to stop open connections.

The tmp file socket might work. I didn't think of that..

Reply:
Using a unix socket in a temp dir works fine and should never conflict, thanks!